### PR TITLE
don't fail spring  application startup if sdk is disabled

### LINKED ISSUE #10600: Disabling OTEL SDK causes autoconfiguration to fail due to missing ConfigProperties
### Describe the bug

With Spring Boot starter, setting `otel.sdk.disabled = true` causes the application startup to fail on on a missing `ConfigProperties` bean dependency. This can be triggered by any of the four autoconfiguration classes that now depend on config properties:
```
OtlpSpanExporterAutoConfiguration
OtlpLogRecordExporterAutoConfiguration
OtlpMetricExporterAutoConfiguration
PropagationAutoConfiguration
```

The issue does not happen in `2.0.0-alpha`, the likely change that triggered it is https://github.com/open-telemetry/opentelemetry-java-instrumentation/commit/f715a189bf65f7de943e6a6b5b5c55cbff531b93, and https://github.com/open-telemetry/opentelemetry-java-instrumentation/commit/d8aa0f5b48e2dee28498b6f95d3d0017eb673cbf for propagation.

### Steps to reproduce

- Create a sample instrumented Spring Boot application as described in https://opentelemetry.io/docs/languages/java/automatic/spring-boot/, with the following configuration:

```yaml
otel:
  sdk:
    disabled: true
```

- Run the application

### Expected behavior

The application starts up, with [SDK entirely disabled](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/autoconfigure/src/main/resources/META-INF/services/io.opentelemetry.sdk.autoconfigure.spi.ResourceProvider#disabling-opentelemetrysdk).

### Actual behavior

The application startup fails with:
```
org.springframework.beans.factory.UnsatisfiedDependencyException:
Error creating bean with name 'otelOtlpLogRecordExporter' defined in class path resource [io/opentelemetry/instrumentation/spring/autoconfigure/exporters/otlp/OtlpLogRecordExporterAutoConfiguration.class]:
Unsatisfied dependency expressed through method 'otelOtlpLogRecordExporter' parameter 0:
No qualifying bean of type 'io.opentelemetry.sdk.autoconfigure.spi.ConfigProperties' available:
expected at least 1 bean which qualifies as autowire candidate. Dependency annotations: {}
 ```
 
 The specific exporter may vary, this is one of 4 possible unsatisfied dependencies.

### Javaagent or library instrumentation version

2.1.0-alpha

### Environment

JDK: Corretto 21
OS: Mac OS Ventura
Spring Boot 3.2.2

### Additional context

As a possible workaround, the application starts up if all exporters and propagation is explicitly disabled:
```yaml
otel:
  sdk:
    disabled: true
  traces:
    exporter: none
  logs:
    exporter: none
  metrics:
    exporter: none
  propagation:
    enabled: false
```

### PR DESCRIPTION
**Pull request by zeitlinger**
Fixes #10600

### CODE REVIEW

**Review by jeanbisutti — COMMENTED** [review]
I woud disable all the autoconfigurations if `otel.sdk.disabled=true`: https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/spring/spring-boot-autoconfigure/src/main/resources/META-INF/spring/org.springframework.boot.autoconfigure.AutoConfiguration.imports

**Comment by jeanbisutti** [comment]
`LogbackAppenderApplicationListener` should also be disabled if  `otel.sdk.disabled=true`: https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/f8788dd2bc6253908ecef0f83473a71008146842/instrumentation/spring/spring-boot-autoconfigure/src/main/resources/META-INF/spring.factories#L21

I have created #10612 to do this.

File: instrumentation/spring/spring-boot-autoconfigure/src/main/java/io/opentelemetry/instrumentation/spring/autoconfigure/exporters/internal/ExporterConfigEvaluator.java [hunk 1]
```diff
@@ -19,6 +19,10 @@ private ExporterConfigEvaluator() {}
   public static boolean isExporterEnabled(
       Environment environment, String exportersKey, String wantExporter, boolean defaultValue) {
 
+    if (environment.getProperty("otel.sdk.disabled", Boolean.class, false)) {
```
Comment (jeanbisutti) [inline instrumentation/spring/spring-boot-autoconfigure/src/main/java/io/opentelemetry/instrumentation/spring/autoconfigure/exporters/internal/ExporterConfigEvaluator.java:22]: I would add a comment, something like this: the method is used to enable some Spring autoconfigurations

File: instrumentation/spring/spring-boot-autoconfigure/src/main/java/io/opentelemetry/instrumentation/spring/autoconfigure/exporters/internal/ExporterConfigEvaluator.java [hunk 2]
```diff
@@ -16,9 +16,17 @@ public final class ExporterConfigEvaluator {
...
   private ExporterConfigEvaluator() {}
 
+  /**
+   * Returns whether the exporter is enabled. This method is used for OTLP, logging and zipkin
+   * exporters.
+   */
   public static boolean isExporterEnabled(
       Environment environment, String exportersKey, String wantExporter, boolean defaultValue) {
 
+    if (environment.getProperty("otel.sdk.disabled", Boolean.class, false)) {
```
Comment (zeitlinger) [inline instrumentation/spring/spring-boot-autoconfigure/src/main/java/io/opentelemetry/instrumentation/spring/autoconfigure/exporters/internal/ExporterConfigEvaluator.java:26]:
```suggestion
    // the method is used to enable some Spring autoconfigurations
    if (environment.getProperty("otel.sdk.disabled", Boolean.class, false)) {
```

Comment (laurit) [reply instrumentation/spring/spring-boot-autoconfigure/src/main/java/io/opentelemetry/instrumentation/spring/autoconfigure/exporters/internal/ExporterConfigEvaluator.java:26]: In my opinion the word `some` is too vague, I have no idea what this comment means.

Comment (zeitlinger) [reply instrumentation/spring/spring-boot-autoconfigure/src/main/java/io/opentelemetry/instrumentation/spring/autoconfigure/exporters/internal/ExporterConfigEvaluator.java:26]: @laurit better?

**Comment by jeanbisutti** [comment]
[`OpenTelemetryAutoConfiguration` is already disabled if `otel.sdk.disabled=true`.](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/617d4edeb5c6ea5c78904723b430300d853dde00/instrumentation/spring/spring-boot-autoconfigure/src/main/java/io/opentelemetry/instrumentation/spring/autoconfigure/OpenTelemetryAutoConfiguration.java#L74)  An alternative implementation could be to autoconfigure the other autoconfigurations after `OpenTelemetryAutoConfiguration`. In this way, if  `OpenTelemetryAutoConfiguration` is disabled, the other ones won't be enabled. This implementation woud require less code and avoid several `otel.sdk.disabled` property verifications.

File: instrumentation/spring/spring-boot-autoconfigure/src/main/java/io/opentelemetry/instrumentation/spring/autoconfigure/exporters/internal/ExporterConfigEvaluator.java (hunk 2 above)
Comment (jeanbisutti) [reply instrumentation/spring/spring-boot-autoconfigure/src/main/java/io/opentelemetry/instrumentation/spring/autoconfigure/exporters/internal/ExporterConfigEvaluator.java:26]: Let's remove the comment if it is ambiguous

**Comment by jeanbisutti** [comment]
> [`OpenTelemetryAutoConfiguration` is already disabled if `otel.sdk.disabled=true`.](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/617d4edeb5c6ea5c78904723b430300d853dde00/instrumentation/spring/spring-boot-autoconfigure/src/main/java/io/opentelemetry/instrumentation/spring/autoconfigure/OpenTelemetryAutoConfiguration.java#L74) An alternative implementation could be to autoconfigure the other autoconfigurations after `OpenTelemetryAutoConfiguration`. In this way, if `OpenTelemetryAutoConfiguration` is disabled, the other ones won't be enabled. This implementation woud require less code and avoid several `otel.sdk.disabled` property verifications.

I mal also fin with the current implementation if you think it is more explicit

**Review by jeanbisutti — APPROVED** [review]

**Comment by zeitlinger** [comment]
> > [`OpenTelemetryAutoConfiguration` is already disabled if `otel.sdk.disabled=true`.](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/617d4edeb5c6ea5c78904723b430300d853dde00/instrumentation/spring/spring-boot-autoconfigure/src/main/java/io/opentelemetry/instrumentation/spring/autoconfigure/OpenTelemetryAutoConfiguration.java#L74) An alternative implementation could be to autoconfigure the other autoconfigurations after `OpenTelemetryAutoConfiguration`. In this way, if `OpenTelemetryAutoConfiguration` is disabled, the other ones won't be enabled. This implementation woud require less code and avoid several `otel.sdk.disabled` property verifications.
> 
> I am also fine with the current implementation if you think it is more explicit

this wouldn't work, because the exporter beans are required to create the OpenTelemetry instance in OpenTelemetryAutoConfiguration

File: instrumentation/spring/spring-boot-autoconfigure/src/main/java/io/opentelemetry/instrumentation/spring/autoconfigure/exporters/internal/ExporterConfigEvaluator.java (hunk 2 above)
Comment (laurit) [reply instrumentation/spring/spring-boot-autoconfigure/src/main/java/io/opentelemetry/instrumentation/spring/autoconfigure/exporters/internal/ExporterConfigEvaluator.java:26]: @zeitlinger yes

File: instrumentation/spring/spring-boot-autoconfigure/src/main/java/io/opentelemetry/instrumentation/spring/autoconfigure/instrumentation/annotations/InstrumentationAnnotationsAutoConfiguration.java [hunk 3]
```diff
@@ -8,20 +8,35 @@
...
 
 /** Configures {@link WithSpanAspect} to trace bean methods annotated with {@link WithSpan}. */
 @ConditionalOnBean(OpenTelemetry.class)
 @ConditionalOnClass(Aspect.class)
-@ConditionalOnProperty(name = "otel.instrumentation.annotations.enabled", matchIfMissing = true)
+@Conditional(InstrumentationAnnotationsAutoConfiguration.Condition.class)
 @Configuration
 public class InstrumentationAnnotationsAutoConfiguration {
+
+  static final class Condition extends AllNestedConditions {
```
Comment (laurit) [inline instrumentation/spring/spring-boot-autoconfigure/src/main/java/io/opentelemetry/instrumentation/spring/autoconfigure/instrumentation/annotations/InstrumentationAnnotationsAutoConfiguration.java:28]: Just curious why you implemented it this way. Would having
```
@ConditionalOnProperty(name = "otel.instrumentation.annotations.enabled", matchIfMissing = true)
@Conditional(SdkEnabled.class)
```
where `SdkEnabled` is shared by all auto configuration classes also work?

Comment (zeitlinger) [reply instrumentation/spring/spring-boot-autoconfigure/src/main/java/io/opentelemetry/instrumentation/spring/autoconfigure/instrumentation/annotations/InstrumentationAnnotationsAutoConfiguration.java:28]: Interesting idea! I hadn't considered that - I'll try.

**Review by laurit — APPROVED** [review]

File: instrumentation/spring/spring-boot-autoconfigure/src/main/java/io/opentelemetry/instrumentation/spring/autoconfigure/exporters/internal/ExporterConfigEvaluator.java (hunk 1 above)
Comment (trask) [reply instrumentation/spring/spring-boot-autoconfigure/src/main/java/io/opentelemetry/instrumentation/spring/autoconfigure/exporters/internal/ExporterConfigEvaluator.java:26]: putting this condition on each exporter's `@Configuration` could make it more visible and consistent (where we just throw the `@SdkEnabled` condition on everything)

Comment (zeitlinger) [reply instrumentation/spring/spring-boot-autoconfigure/src/main/java/io/opentelemetry/instrumentation/spring/autoconfigure/exporters/internal/ExporterConfigEvaluator.java]: great idea - done :smile:

Comment (jeanbisutti) [reply instrumentation/spring/spring-boot-autoconfigure/src/main/java/io/opentelemetry/instrumentation/spring/autoconfigure/exporters/internal/ExporterConfigEvaluator.java]: > putting this condition on each exporter's `@Configuration` could make it more visible and consistent (where we just throw the `@SdkEnabled` condition on everything)

Should it not be another property configuration allowing to disable the export? 

By setting `otel.sdk.disabled=true`, I understand disable the OpenTelemtry SDK, and so disable  the instrumentation features and the export.

Comment (zeitlinger) [reply instrumentation/spring/spring-boot-autoconfigure/src/main/java/io/opentelemetry/instrumentation/spring/autoconfigure/exporters/internal/ExporterConfigEvaluator.java]: > > putting this condition on each exporter's `@Configuration` could make it more visible and consistent (where we just throw the `@SdkEnabled` condition on everything)
> 
> Should it not be another property configuration allowing to disable the export?

you still have `otel.traces.exporter=none` (similar for logs and metrics)

> By setting `otel.sdk.disabled=true`, I understand disable the OpenTelemtry SDK, and so disable the instrumentation features and the export.

yes, `otel.sdk.disabled=true` disables everything at once.

**Pull request merged**